cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/FileSystemStorageAdapter.java

```diff
@@ -0,0 +1,523 @@
+        StringBuilder path = new StringBuilder();
+        int bucketDepth = 0;
+        String fixedParentDir = null;
+        switch (bucketMode) {
+            case URI_BASED:
+                if (bucket != null && bucket.length() > 0) {
+                    try {
+                        URI test = new URI(bucket + "/file");
+                        InventoryUtil.validateArtifactURI(FileSystemStorageAdapter.class, test);
+                    } catch (URISyntaxException | IllegalArgumentException e) {
+                        throw new IllegalArgumentException("bucket must be in the form 'scheme:path'");
+                    }
+                    int colonIndex = bucket.indexOf(":");
+                    path.append(bucket.substring(0, colonIndex + 1));
+                    path.append(File.separator);
+                    path.append(bucket.substring(colonIndex + 1));
+                }
+                if (path.length() > 0) {
+                    fixedParentDir = path.toString();
+                }
+                break;
+            case URI_BUCKET_BASED:
+                if (bucket != null) {
+                    if (bucket.length() > BUCKET_LENGTH) {
+                        throw new IllegalArgumentException("bucket must be a maximum of " + BUCKET_LENGTH + " characters");
+                    }
+                    for (char c : bucket.toCharArray()) {
+                        path.append(c).append(File.separator);
+                    }
+                    bucketDepth = BUCKET_LENGTH - bucket.length();
+                } else {
+                    bucketDepth = BUCKET_LENGTH;
+                }
+                break;
+            default:
+                throw new IllegalStateException("unsupported bucket mode");
+        }
+        try {
+            log.debug("resolving path: " + path.toString());
+            Path bucketPath = root.resolve(path.toString());
+            if (!Files.exists(bucketPath) || !Files.isDirectory(bucketPath)) {
+                throw new IllegalArgumentException("Invalid bucket: " + bucket);
+            }
+            Iterator<StorageMetadata> iterator = null;
+            switch (bucketMode) {
+                case URI_BASED:
+                    iterator = new ArtifactIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                case URI_BUCKET_BASED:
+                    iterator = new FileSystemIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                default:
+                    throw new IllegalStateException("unsupported bucket mode");
+            }
+            return iterator;
+        } catch (InvalidPathException e) {
+            throw new IllegalArgumentException("Invalid bucket: " + bucket);
+        }
+    }
+    
+    private Path createArtifactPath(URI artifactURI, String uriBucket) {
+        StringBuilder path = new StringBuilder();
+        switch (bucketMode) {
+            case URI_BASED:
```

Comment:
Yes, the directory structure in this mode for an artifactURI of cadc:TEST/my/file would be fsroot/cadc:/TEST/my/file.  Bucketing in this mode can be any depth of directory, so you could bucket by cadc: or cadc:TEST or cadc:TEST/my.

However, note how there are no slashes after the scheme in these bucket examples.  I decided to let the bucket resemble the artifactURI substrings rather than the storageURI substrings for usability.  So, cadc:TEST rather than cadc:/TEST.